### PR TITLE
Basic string formatting

### PR DESCRIPTION
**Pull request by MichaReiser**
## Summary

This PR implements formatting for non-f-string Strings that do not use implicit concatenation. 

Docstring formatting is out of the scope of this PR.

## Test Plan

I added a few tests for simple string literals. 

## Performance

Ouch. This is hitting performance somewhat hard. This is probably because we now iterate each string a couple of times:

1. To detect if it is an implicit string continuation
2. To detect if the string contains any new lines
3. To detect the preferred quote
4. To normalize the string

Edit: I integrated the detection of newlines into the preferred quote detection so that we only iterate the string three time.
We can probably do better by merging the implicit string continuation with the quote detection and new line detection by iterating till the end of the string part and returning the offset. We then use our simple tokenizer to skip over any comments or whitespace until we find the first non trivia token. From there we keep continue doing this in a loop until we reach the end o the string. I'll leave this improvement for later.

### CODE REVIEW

**Comment by MichaReiser** [comment]
Current dependencies on/for this PR:
* main
  * **PR #5289** <a href="https://app.graphite.dev/github/pr/astral-sh/ruff/5289" target="_blank">Graphite</a> 
    * **PR #5292** <a href="https://app.graphite.dev/github/pr/astral-sh/ruff/5292" target="_blank">Graphite</a> 
      * **PR #5297** <a href="https://app.graphite.dev/github/pr/astral-sh/ruff/5297" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/astral-sh/ruff/5297?utm_source=stack-comment).

File: crates/ruff_python_formatter/src/expression/string.rs [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+        }
+
+        // Drop unicode
+
+        Ok(())
+    }
+}
+
+/// Detects the preferred quotes for `input`. The preferred quote style is the one that
+/// requires less escape sequences.
```
Comment (konstin) [inline crates/ruff_python_formatter/src/expression/string.rs:147]: i didn't realize this when using black, i always thought it was doing double quotes always, interesting to know

File: crates/ruff_python_formatter/src/expression/string.rs [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+    let mut chars = input.char_indices().peekable();
+
+    while let Some((index, c)) = chars.next() {
+        if c == '\\' {
+            if let Some((_, next)) = chars.peek() {
+                if *next == opposite_quote {
+                    chars.next();
+                    // Remove the escape
+                    output.push_str(&input[last_index..index]);
+                    last_index = index + '\\'.len_utf8();
```
Comment (konstin) [inline crates/ruff_python_formatter/src/expression/string.rs:235]:
```suggestion
                    // Remove the escape by ending before the backstring and starting again with the quote
                    last_index = index + '\\'.len_utf8();
```

File: crates/ruff_python_formatter/src/expression/string.rs [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+            output.push_str(&input[last_index..index]);
+            output.push('\\');
+            output.push(c);
+            last_index = index + preferred_quote.len_utf8();
+        } else if matches!(c, '\n' | '\r') {
+            contains_newlines = ContainsNewlines::Yes;
+        }
+    }
+
+    let normalized = if last_index == 0 {
```
Comment (konstin) [inline crates/ruff_python_formatter/src/expression/string.rs:252]: that's a neat trick, but i think it needs a comment somewhere at the top of the function

File: crates/ruff_python_formatter/src/lib.rs [hunk 4]
```diff
@@ -342,29 +377,8 @@ if True:
         let printed = format_module(&content)?;
         let formatted_code = printed.as_code();
 
-        let reformatted =
```
Comment (konstin) [inline crates/ruff_python_formatter/src/lib.rs:345]: why is that gone?

File: crates/ruff_python_formatter/src/expression/string.rs [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+        TextSize::new(self.bits().count_ones())
+    }
+}
+
+impl Format<PyFormatContext<'_>> for StringPrefix {
+    fn fmt(&self, f: &mut Formatter<PyFormatContext<'_>>) -> FormatResult<()> {
+        if self.contains(StringPrefix::RAW) {
+            text("r").fmt(f)?;
+        } else if self.contains(StringPrefix::RAW_UPPER) {
+            text("R").fmt(f)?;
```
[the remark this answers is not in the document]
Comment (MichaReiser) [reply crates/ruff_python_formatter/src/expression/string.rs:129]: I should probably add a comment linking to https://black.readthedocs.io/en/stable/the_black_code_style/current_style.html#r-strings-and-r-strings

File: crates/ruff_python_formatter/src/expression/string.rs [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+/// Detects the preferred quotes for `input`. The preferred quote style is the one that
+/// requires less escape sequences.
+fn preferred_quotes(input: &str) -> QuoteStyle {
+    let mut single_quotes = 0u32;
+    let mut double_quotes = 0u32;
+    let mut chars = input.chars();
+
+    while let Some(c) = chars.next() {
+        let style = match c {
+            '\\' => chars.next().ok_or(()).and_then(QuoteStyle::try_from),
```
[the remark this answers is not in the document]
Comment (MichaReiser) [reply crates/ruff_python_formatter/src/expression/string.rs:155]: I simplified the logic a bit. What this code does is counting the `'` and `"` but in a single pass. Calling `.count()` would (possibly) traverse the string twice.

**Comment by MichaReiser** [comment]
[Graphite](https://app.graphite.dev/github/pr/astral-sh/ruff/5297) rebased this pull request after merging its parent, because this pull request is set to merge when ready.

**Comment by MichaReiser** [comment]
@MichaReiser merged this pull request with [Graphite](https://app.graphite.dev/github/pr/astral-sh/ruff/5297).

**Pull request merged**